### PR TITLE
Bytecode, and a collection of bytecodes.

### PR DESCRIPTION
**Pull request by ChihChengLiang**
### Description

We revisit the handling of the bytecodes in the codebase. We reduce manual manipulation of bytes and use a more streamlined API with eth_types::Bytecode and bus_mapping::CodeDB.

### Issue Link

Missed out work in #1391 

### Type of change

Refactor

### Contents

- Removed witness::bytecode module.
- Extend eth_types::Bytecode and bus_mapping::CodeDB.
  - eth_types::Bytecode handles a single bytecode instance, and
  - CodeDB handles a group of bytecode instances
- rewrite the bytecode circuit witness assignment.
- rewrite the bytecode circuit tests, so that overwriting will not be included in the production code.
- deduplicated the Mock account testing code.

### Rationale

- I kept the scope of the change small to reduce the review burden.
- I will add more rationale with review comments.

### CODE REVIEW

File: bus-mapping/src/state_db.rs
Comment (ChihChengLiang) [inline bus-mapping/src/state_db.rs:67]: I feel this is a bad API. We shouldn't expose raw bytes for people to manipulate.
It is currently used in two places.
- One is in the keccak circuit, which takes `Vec<Vec<u8>>`. I didn't address keccak circuit in this pr to limit the scope.
- The other one is BytecodeCircuitAssignment. We need the raw bytes to preserve bytecode order for the test cases.

File: eth-types/src/bytecode.rs
Comment (ChihChengLiang) [inline eth-types/src/bytecode.rs:199]: I keep the original code from tracer tests. But I keep wondering if we should pad only zeros but not `0...something`

File: eth-types/src/geth_types.rs
Comment (ChihChengLiang) [inline eth-types/src/geth_types.rs:68]: I feel these two APIs shouldn't be here. They are here to de-duplicate the test codes. Let me know if you have thoughts on this.

File: testool/src/statetest/executor.rs
Comment (ChihChengLiang) [inline testool/src/statetest/executor.rs:83]: I don't get why we currently need `std::borrow::Cow::Owned` and `std::borrow::Cow::Borrowed`

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ChihChengLiang) [inline zkevm-circuits/src/bytecode_circuit/circuit.rs:490]: I removed fail_fast because it was never triggered.

File: bus-mapping/src/state_db.rs
Comment (hero78119) [reply bus-mapping/src/state_db.rs:67]: I think there is option also avoiding refactor pretty much for other part is implementing iterator trait on CodeDB and iterate through &byte (need to check whether at the caller side, still need byte copy or not)

File: bus-mapping/src/state_db.rs
Comment (hero78119) [inline bus-mapping/src/state_db.rs:25]: Since codedb is a heavy object, are we able to remove `Clone` trait for it

**Review by hero78119 — COMMENTED** [review]
Just have a first pass, all changes outside bytecode circuit LGTM! thanks for using new api to clean up codebase quite a lot! 

Will have second review on bytecode circuit logic thoroughly sooner.

File: eth-types/src/geth_types.rs
Comment (ChihChengLiang) [reply eth-types/src/geth_types.rs:68]: The `Account` struct is located at the eth-types crate. But the flat "test-circuits" is in the zkevm-circuits crate.
Adding a flag to `eth-types` would complicate compilation and save only a little compilation time. 
I'm thinking of cleaning up the MockAccount in the mock crate next time and moving these two methods there.

File: bus-mapping/src/state_db.rs
Comment (ChihChengLiang) [reply bus-mapping/src/state_db.rs]: I realized CodeDB has into_iterator trait. I use it to replace to_raw in [bcbef94](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/bcbef9417b78d3245b3d41de75865d105c6ae699)

File: testool/src/statetest/executor.rs
Comment (ChihChengLiang) [reply testool/src/statetest/executor.rs]: I discussed this with @ed255. We concluded that the Cow was there to prevent cloning twice.

- Cow is used to keep `actual_code` a reference to the byte vector so that we don't copy bytecodes from codeDB here.
- The line `found: Bytes::from(actual_code.to_vec()),` below clones bytes with the `to_vec` method.

I resolved with another method that only clones once.

File: zkevm-circuits/src/bytecode_circuit/circuit.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/bytecode_circuit/circuit.rs:119]: I checked with @ed255 and concluded that `cur.push_data_left` is not underconstrained.

When current tag is "Header" and the next is "Byte", the next "is_code" will be constrained as 1, which indirectly constrains the `cur.push_data_left` of the first "Byte" row to be 0.

File: bus-mapping/src/state_db.rs
Comment (ChihChengLiang) [reply bus-mapping/src/state_db.rs:25]: It would be hard to remove `Clone` right now. Many other objects like `BlockData` and `Block` depend on it.

**Review by hero78119 — APPROVED** [review]
LGTM! The rewrite of bytecode witness assignment and unify padding/normal row into single set_row api is awesome, plus the bytecode test api reorganized. 

Just have few nitpick regarding to function naming, and return type suggestion.

**Review by ed255 — COMMENTED** [review]
Overall LGTM!  Very nice clean up :tada:
I left 3 comments, please take a look!

File: bus-mapping/src/state_db.rs
Comment (ChihChengLiang) [reply bus-mapping/src/state_db.rs:59]: Good point. I fixed them in this commit [rename to get_from_u256 and get_from_h256](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/4d5ade22d90c97adc664abb8a71ee804bf2d97c7)

File: bus-mapping/src/state_db.rs
Comment (ChihChengLiang) [reply bus-mapping/src/state_db.rs]: Yes, it turns out that only a few places in the codebase use this API. I've removed it in [dc2a85b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/dc2a85be0c5bcd268593782f3d164981aeb120cc)

File: zkevm-circuits/src/bytecode_circuit/test.rs
Comment (ChihChengLiang) [reply zkevm-circuits/src/bytecode_circuit/test.rs]: I find no good reason to have that +3, so I've removed it. [6b5574b](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/6b5574b4088e6af0203ef7eb4305e1395de09c98)

File: eth-types/src/bytecode.rs
Comment (ChihChengLiang) [reply eth-types/src/bytecode.rs]: I've doc this and fix the padding zeros in [3abae63](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/3abae63164b50f50dc986a2c7735e149891fa433)

File: eth-types/src/geth_types.rs
Comment (ChihChengLiang) [reply eth-types/src/geth_types.rs:55]: Nice catch! I renamed the function and updated the documentation [2228e4d](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1529/commits/2228e4deddf92d028661343e05c4af08285eda04)

**Review by ed255 — APPROVED** [review]
LGTM!  Great clean up :dancing_men:

**Pull request merged**
